pyllamacpp/model.py

```diff
@@ -137,19 +155,11 @@ def generate(self,
         :param verbose: if `True`, `llama.cpp` stuff will be printed
         :return: Tokens generator
         """
-        prompt = f' {self.prompt_prefix}{prompt}{self.prompt_suffix}'
-        input_tokens = pp.llama_tokenize(self._ctx, prompt, True)
-        if len(input_tokens) > self._n_ctx - 4:
+        input_tokens = self._prompt_prefix_tokens + pp.llama_tokenize(self._ctx, prompt, True) + self._prompt_suffix_tokens
+        if len(input_tokens) > self._n_ctx - 4 - len(self._prompt_context_tokens):
             raise Exception('Prompt too long!')
-        predicted_tokens = []
-        predicted_token = 0
-
-        # add global context for the first time
-        if self._n_past == 0:
-            for tok in self._prompt_context_tokens:
-                predicted_tokens.append(tok)
-                self._last_n_tokens.pop(0)
-                self._last_n_tokens.append(tok)
+        predicted_tokens = self._prompt_context_tokens.copy()
```

Comment:
What this variable will be used for ?

Reply:
The .copy() is crucial as I don't want to change the content of self._prompt_context_tokens.
I need the context to continue existing in the text all the time so that the bot doesn't forget who he is.
In previous code, you only put the context one time then let the buffer slide by pushing new stuff and poping old stuff. But when you reach context size, you loose the context conditionning (what you call context_tokens and I would rather call conditionning tokens).

In my code, the model is applyed on a constructed buffer that contains the context tokens as fixed thing, and the window slides on the rest of the text (the _last_n_tokens variable).

For example, at the beginning
your prompt is abcdefgh....stuff
The conditioning text is: <Conditioning text>
The prefixtext is: ##prefix\n
The suffixtext is:  ##suffix\n

Your sliding window looks like:
_last_n_tokens => ##prefix\nabcdefgh....stuff##suffix\n

Your reconstructed prompt is :
input_prompt = self._prompt_context_tokens + self._last_n_tokens[self._n_ctx - self.n_consumed_tokens:]

<Conditioning text>##prefix\nabcdefgh....stuff##suffix\n

the model answers jjjj, the _last_n_tokens  should look like:
##prefix\nabcdefgh....stuff##suffix\njjjj

then the user types blabla, so the new _last_n_tokens  looks tile:
##prefix\nabcdefgh....stuff##suffix\njjjj##prefix\nblabla##suffix\n

Your reconstructed prompt is :
<Conditioning text>##prefix\nabcdefgh....stuff##suffix\njjjj##prefix\nblabla##suffix\n

Now you continue doing this until your _last_n_tokens reaches its maximum and you start to loose tokens at the beginning. Using my code, you will always continue to have the conditionning in your context as you roll on the rest but the context text stays untouched 

Your reconstructed prompt is :
<Conditioning text>defgh....stuff##suffix\njjjj##prefix\nblabla##suffix\nnew long text


This ensures that we don't loose conditionning even if the whole context size is consumed.

Reply:
Ok, I understand,

But that comes with performance issue, as the conditioning is "usually" a long sentence, and adding it on every function call will make the function slower !! I think we should leave the user decide what they should do depending on their case. 

What do you think ?